kernel/posix/semaphore.c

```diff
@@ -136,3 +139,342 @@ int sem_wait(sem_t *semaphore)
 	k_sem_take(semaphore, K_FOREVER);
 	return 0;
 }
+
+
+/*
+ * Named semaphore design:
+ * - Opened table -> which semaphore is open by which thread.
+ * - Named semaphore pool -> global struct to allocate named sem.
+ */
+
+/*
+ * Used for error setting and exit
+ */
+#define SEM_ERRNO_SET(_code)      \
+	do {                      \
+		errno = (_code);  \
+		goto error;       \
+	} while (0)
+
+#define NSEM_SYS_N_MAX     (32) /* Limit of allocated nsem in system */
+#define NSEM_SYS_OPEN_MAX  (32) /* System limit of simultaneously open nsem */
+#define NSEM_PROC_OPEN_MAX  (4) /* Process limit of simultaneously open nsem */
+
+/*
+ * Private named semaphore
+ */
+struct _nsem {
+	sem_t sem;
+	const char *name;
+	bool to_free; /* Flag set by unlink */
+	bool in_use;  /* Used to check if this spot is available */
+};
+static unsigned int sem_named_n;
+static struct _nsem sem_named[NSEM_SYS_N_MAX];
+
+/*
+ * Given a sem_t value, return the _nsem object containing it.
+ * This needs to be updated if we change the order of the elements in
+ * struct _nsem.
+ */
+#define nsem_container_of(sem_t_ptr) ((struct _nsem *)sem_t_ptr)
```

Comment:
There is already a more robust (no need to update it when the field moves in the parent struct) CONTAINER_OF macro in misc/util.h that does what you want here.

Reply:
I grepped for the lowercase one, so couldn't find it. Thanks for pointing out.